### PR TITLE
feat(track) track call improvements for LP2.

### PR DESCRIPTION
**Pull request by Enuviel**

### CODE REVIEW

**Comment by alexisoyama** [comment]
test this please

**Comment by Enuviel** [comment]
test this please

File: Leanplum-SDK/Classes/Leanplum.m [hunk 1]
```diff
@@ -3519,48 +3521,6 @@ - (NSDictionary *)getChildArgs:(NSString *)name
     LP_END_TRY
 }
 
-/**
- * Prefix given event with all parent actionContext names to while filtering out the string
- * "action" (used in ExperimentVariable names but filtered out from event names).
- */
-- (NSString *)eventWithParentEventNamesFromEvent:(NSString *)event
```
Comment (alexisoyama) [inline Leanplum-SDK/Classes/Leanplum.m]: Let's keep this. It is easier to understand why we have it. "Used in ExperimentVariable" etc..

File: Leanplum-SDK/Classes/Leanplum.m [hunk 2]
```diff
@@ -3620,14 +3591,68 @@ - (void)trackMessageEvent:(NSString *)event withValue:(double)value andInfo:(NSS
...
+                                                               withString:@""
+                                                                  options:NSCaseInsensitiveSearch
+                                                                    range:NSMakeRange(0,
+                                                                                      actionName.length)
+                          ];
+            if (fullEventName.length) {
+                [fullEventName appendString:@" "];
+            }
+            [fullEventName appendString:actionName];
+            for(NSString *messageKey in messagesForTrack.allKeys){
```
Comment (alexisoyama) [inline Leanplum-SDK/Classes/Leanplum.m]: Let's make a new method just for this similar to  eventWithParentEventNamesFromEvent:
Something like
- (NSDictionary *)messagesToTrackFromEvent:(NSString *)event;
Have a clear documentation on when this method will be used: Chained message for Accept Accept and such.

Have another method
- (void)enumerateParentContextWithBlock:(void (^)(LPActionContext *context, NSInteger index, NSString *actionName));
That will be used both by messagesToTrackFromEvent: and messagesToTrackFromEvent:

Comment (Enuviel) [inline Leanplum-SDK/Classes/Leanplum.m]: how enumerateParentContextWithBlock will looks?

File: Leanplum-SDK/Classes/Leanplum.m [hunk 3]
```diff
@@ -3582,8 +3542,19 @@ - (void)runActionNamed:(NSString *)name
             LPActionContext *chainedActionContext =
                     [Leanplum createActionContextForMessageId:messageId];
             chainedActionContext->_parentContext = self;
+            chainedActionContext.contextualValues = self.contextualValues;
+            chainedActionContext->_preventRealtimeUpdating = _preventRealtimeUpdating;
+            chainedActionContext->_isRooted = _isRooted;
+            chainedActionContext->_parentContext = self;
+            chainedActionContext->_key = name;
+            chainedActionContext->_chainedToExisting = YES;
             dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (Enuviel) [inline Leanplum-SDK/Classes/Leanplum.m]: done

File: Leanplum-SDK/Classes/Leanplum.m [hunk 4]
```diff
@@ -3143,7 +3144,7 @@ + (LPActionContext *)actionContextWithName:(NSString *)name
 }
 
 + (LPActionContext *)actionContextWithName:(NSString *)name
-                                      args:(NSDictionary *)args
+                                      args:(NSDictionary *)args
```
Comment (benmarten) [inline Leanplum-SDK/Classes/Leanplum.m]: remove new space?

File: Leanplum-SDK/Classes/Leanplum.m (hunk 2 above)
Comment (alexisoyama) [inline Leanplum-SDK/Classes/Leanplum.m]: `- (void)enumerateParentContextWithBlock:(void (^)(LPActionContext *context, NSInteger index, NSString *actionName)block) {
         LPActionContext *context = self;
         NSMutableArray *parents = [NSMutableArray array];
         while (context->_parentContext != nil) {
             [parents addObject:context];
             context = context->_parentContext;
         }

         NSString *actionName;
         NSMutableDictionary *messagesForTrack= [[NSMutableDictionary alloc] init];
        for (NSInteger i = parents.count - 1; i >= -1; i--) {
            NSString *messageIdForTrack = _messageId;
             context=self;
             if (i > -1) {
                 context = (LPActionContext *) parents[i];
                 actionName =  context->_key;
             } else {
                 actionName = event;
             }
             if (actionName == nil) {
                 break;
             }

             actionName = [actionName stringByReplacingOccurrencesOfString:@" action"
                                                                withString:@""
                                                                   options:NSCaseInsensitiveSearch
                                                                     range:NSMakeRange(0,
                                                                                       actionName.length)
                           ];
               block(context, i, actionName);
     }
}`

Comment (alexisoyama) [inline Leanplum-SDK/Classes/Leanplum.m]: Damn this is hard to read. Will slack you.

File: Leanplum-SDK/Classes/Leanplum.m [hunk 5]
```diff
@@ -3620,14 +3592,69 @@ - (void)trackMessageEvent:(NSString *)event withValue:(double)value andInfo:(NSS
...
+                messagesForTrack[messageIdForTrack] = actionName;
+            }
+        }
+        if (!_chainedToExisting && fullEventName) {
+            [Leanplum track:fullEventName
                   withValue:value
                     andInfo:info
                     andArgs:@{LP_PARAM_MESSAGE_ID: _messageId}
               andParameters:params];
+        }else if (_chainedToExisting) {
```
Comment (benmarten) [inline Leanplum-SDK/Classes/Leanplum.m]: space: }else

File: Leanplum-SDK/Classes/Leanplum.m [hunk 6]
```diff
@@ -3620,14 +3592,69 @@ - (void)trackMessageEvent:(NSString *)event withValue:(double)value andInfo:(NSS
...
+        NSString *actionName;
+        NSMutableDictionary *messagesForTrack= [[NSMutableDictionary alloc] init];
+        
+        for (NSInteger i = parents.count - 1; i >= -1; i--) {
+            NSString *messageIdForTrack = _messageId;
+            context=self;
+            if (i > -1) {
+                context=(LPActionContext *) parents[i];
+                actionName =  context->_key;
+                if(context->_chainedToExisting){
```
Comment (benmarten) [inline Leanplum-SDK/Classes/Leanplum.m]: space

**Comment by Enuviel** [comment]
As discussed yesterday, we will have to each message only two events:  view and Accept.

**Pull request merged**